### PR TITLE
Mccalluc/antibodies

### LINKED ISSUE #431: Create antibody_metadata schema
Here is a link to the template:
https://docs.google.com/spreadsheets/d/1ZzwL_-Ckj9jjQk2CD9utFcZmQFIqFkuUkuJQxVhAVq8/edit?usp=sharing

### PR DESCRIPTION
**Pull request by mccalluc**
Fix #431.
@cebriggs7135 :
- Please make inline suggestions to fill out the descriptions in the YAML.
- I was thinking this would be referenced from the dataset metadata.tsv... Does that match your understanding?
- What assay types is this applicable to?

### CODE REVIEW

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+fields:
+  # NOTE: Example had "dataset_uuid": Won't work. Not known at submission time.
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:2]: How we could associate antibody metadata on each row of an antibody.tsv to the relevant dataset before a uuid is generated?

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    name: uniprot_accession_number
+    description: TODO
+    # TODO: pattern constraint
+  -
+    name: lot_number
+    description: TODO
+  -
+    name: validation_report_location
+    description: TODO
+    # Is this part of the submission? Or a URL? If a URL, put that in the field name.
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:20]: This is what we decided on while waiting for Bill Shirey to confirm:  users should create a directory called "extras" and put this Ab tsv and the Ab validation report in it. Would the proper field entry then be "/Ab_validation_report_123xyz.pdf" ?

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+fields:
+  # NOTE: Example had "dataset_uuid": Won't work. Not known at submission time.
+  -
+    name: antibody_name
+    description: TODO
+  -
+    name: rrr_id
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:8]:
```suggestion
    description: The rrr_id is a unique antibody identifier (eg: AB_#######) that comes from the Antibody Registry (https://antibodyregistry.org).
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+fields:
+  # NOTE: Example had "dataset_uuid": Won't work. Not known at submission time.
+  -
+    name: antibody_name
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:5]:
```suggestion
    description: Anti-(target name) antibody.
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  -
+    name: rrr_id
+    description: TODO
+    # TODO: pattern constraint
+  -
+    name: uniprot_accession_number
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:12]:
```suggestion
    description: The uniprot_accession_number is a unique identifier (eg: P#####) for proteins in the UniProt database (https://www.uniprot.org).
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  -
+    name: lot_number
+    description: TODO
+  -
+    name: validation_report_location
+    description: TODO
+    # Is this part of the submission? Or a URL? If a URL, put that in the field name.
+  -
+    name: assay
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:23]:
```suggestion
    description: This antibody.tsv is unique to a specific HuBMAP assay dataset. Enter the name of the specific assay using this antibody preparation in this field. If the antibody prep was used for multiple assays, the metadata should be re-entered on one row per assay.
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  -
+    name: dilution
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:26]:
```suggestion
    description: Antibody solutions may be diluted according to the experimental protocol.
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    # TODO: pattern constraint?
+  -
+    name: conjugated_cat_number
+    description: TODO
+  -
+    name: conjugated_tag
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:33]:
```suggestion
    description: The name of the entity conjugated to the antibody. (optional) .
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  -
+    name: assay
+    description: TODO
+  -
+    name: dilution
+    description: TODO
+    # TODO: pattern constraint?
+  -
+    name: conjugated_cat_number
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:30]:
```suggestion
    description: The catalog number for conjugated antibody. (optional) 
```

**Review by cebriggs7135 — CHANGES_REQUESTED** [review]

**Comment by cebriggs7135** [comment]
@mccalluc 

- I was thinking this would be referenced from the dataset metadata.tsv... Does that match your understanding?
Yes. This is the suggested mechanism for now: Users should create a directory called "extras" and put this Ab tsv and the Ab validation report in it.
- What assay types is this applicable to?
CODEX, MxIF, any other assays with "immuno-" in the name, which indicates use of antibodies.

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    name: rrr_id
+    description: TODO
+    # TODO: pattern constraint
+  -
+    name: uniprot_accession_number
+    description: TODO
+    # TODO: pattern constraint
+  -
+    name: lot_number
+    description: TODO
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:16]:
```suggestion
    description: The lot# is specific to the vendor. (eg: Abcam lot# GR3238979-1)
```

**Review by cebriggs7135 — CHANGES_REQUESTED** [review]

File: src/ingest_validation_tools/table-schemas/antibodies.yaml (hunk 1 above)
Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml:2]: >> I was thinking this would be referenced from the dataset metadata.tsv... Does that match your understanding?

> Yes. This is the suggested mechanism for now: Users should create a directory called "extras" and put this Ab tsv and the Ab validation report in it.

From the quotation above, I was thinking that each dataset would reference an entire antibody list... Is it clear to you how that is different from having the antibodies rows reference datasets?

Let me finish integrating the other changes here, and then let's make time to talk this through. Probably hard to resolve messaging back and forth.

File: src/ingest_validation_tools/table-schemas/antibodies.yaml (hunk 2 above)
Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml:20]: Ok, let's call this `validation_report_path` then, and if its in `extras`, the value would be `extras/ab-validation-report.pdf` or something like that. (Let's not use mixed case in filenames unless there's a reason.)

File: src/ingest_validation_tools/table-schemas/antibodies.yaml (hunk 6 above)
Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml:23]: This goes to the data-modelling issue... Need to talk/clarify.

Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml:27]: @cebriggs7135 : Can you rephrase this as a description of what the field means? We want a description that could be used downstream... instructions for how to fill it in are not what the data consumer needs.

**Comment by mccalluc** [comment]
Still TODO:
- add reference columns from applicable metadata TSVs
- add a test fixture

Good to show this to Jonathan when we meet, but perhaps not needed before merge.

**Comment by mccalluc** [comment]
> > What assay types is this applicable to?

> CODEX, MxIF, any other assays with "immuno-" in the name, which indicates use of antibodies.

@cebriggs7135 : Looking at `assay_type`, there are none with "immuno-" in the name, so I'm afraid I'm missing something.

File: src/ingest_validation_tools/table-schemas/antibodies.yaml (hunk 6 above)
Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml]: moot

File: src/ingest_validation_tools/table-schemas/antibodies.yaml (hunk 8 above)
Comment (mccalluc) [reply src/ingest_validation_tools/table-schemas/antibodies.yaml]: done

**Comment by mccalluc** [comment]
Important parts to look at:
- [Revised submission structure and diagram](https://github.com/hubmapconsortium/ingest-validation-tools/blob/baa49b0f85375c3a83a3822f7419f478ee672795/README.md#submission-directory-model)
- [Antibodies TSV docs](https://github.com/hubmapconsortium/ingest-validation-tools/blob/baa49b0f85375c3a83a3822f7419f478ee672795/docs/antibodies/README.md#antibodies)

I think this could be merged as-is, but there is remaining work:
- Actually validate contributors.tsv and add to tests: #434
- Notes about optional fields in antibodies.tsv are not being generated: #435
- Identify other "immuno" assay types: #436

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 11]
```diff
@@ -0,0 +1,38 @@
+fields:
+  -
+    name: channel_id
+    description: The cycle/channel in which the protein was assessed with this antibody.
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:4]:
```suggestion
    description: The cycle/channel in which the protein was assessed with this antibody. For example, an experiment that involves 3 hybridization cycles & 4 signal detection channels would have **channel_id**s numbered 1-12. Antibodies detected with channel 1 would have **channel_id** 1,5 & 9. Antibodies detected with channel 2 would have **channel_id** 2,6 & 10 and so on for the remaining antibodies.
```

File: src/ingest_validation_tools/table-schemas/antibodies.yaml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  -
+    name: dilution
+    description: Antibody solutions may be diluted according to the experimental protocol.
+    constraints:
+      pattern: '1/\d+'
+    example: 1/200
+    required: False
+  -
+    name: conjugated_cat_number
+    description: The catalog number for conjugated antibody.
```
Comment (cebriggs7135) [inline src/ingest_validation_tools/table-schemas/antibodies.yaml:33]:
```suggestion
    description: An antibody may be conjugated to a fluorescent tag or a metal tag for detection. Conjugated antibodies may be purchased from commercial providers. 
```

**Review by cebriggs7135 — CHANGES_REQUESTED** [review]
I've added descriptive text to the fields for conjugated antibodies.

File: docs/antibodies/README.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+Related files:
+
+- [📝 TSV template](https://raw.githubusercontent.com/hubmapconsortium/ingest-validation-tools/master/docs/antibodies/antibodies-metadata.tsv): Use this to submit metadata.
+- [💻 Source code](https://github.com/hubmapconsortium/ingest-validation-tools/edit/master/src/ingest_validation_tools/table-schemas/antibodies.yaml): Make a PR if this doc should be updated.
+
+## Table of contents
+[`channel_id`](#channel_id)<br>
+[`antibody_name`](#antibody_name)<br>
+[`rrr_id`](#rrr_id)<br>
```
Comment (cebriggs7135) [inline docs/antibodies/README.md:11]:
```suggestion
[`rr_id`](#181 rr_id)<br>
```

Turns out Marda says it should be rr_id.

File: docs/antibodies/README.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+| --- | --- |
+| type | `integer` |
+
+### `antibody_name`
+Anti-(target name) antibody. Not validated or used down-stream.
+
+
+
+### `rrr_id`
+The rrr_id is a unique antibody identifier that comes from the Antibody Registry (https://antibodyregistry.org). Example: `AB_10002075`.
```
Comment (cebriggs7135) [inline docs/antibodies/README.md:31]:
```suggestion
The rr_id is a unique antibody identifier that comes from the Antibody Registry (https://antibodyregistry.org). Example: `AB_10002075`.
```

File: docs/antibodies/antibodies.tsv [hunk 15]
```diff
@@ -0,0 +1,2 @@
+channel_id	antibody_name	rrr_id	uniprot_accession_number	lot_number	dilution	conjugated_cat_number	conjugated_tag
```
Comment (cebriggs7135) [inline docs/antibodies/antibodies.tsv:1]:
```suggestion
channel_id	antibody_name	rr_id	uniprot_accession_number	lot_number	dilution	conjugated_cat_number	conjugated_tag
```

**Review by cebriggs7135 — CHANGES_REQUESTED** [review]
rrr_id should be rr_id

**Comment by cebriggs7135** [comment]
@mccalluc 
re: your comment above https://github.com/hubmapconsortium/ingest-validation-tools/pull/432#issuecomment-733927977

"I" in MxIF stands for multiplex immunofluorescence. 

Here are other new assays that use antibodies:

ImmunoSABER
Cell DIVE MxIF (cell DIVE is a commercial platform for hyperplaxed MxIF)
MxIHC (Immuno-Histo-Compatibility)
MIBI
lightsheet

File: docs/antibodies/README.md (hunk 13 above)
Comment (mccalluc) [reply docs/antibodies/README.md:11]: Why 181? I'll correct rrr to rr, but let me know if the number is there for a reason.

**Comment by mccalluc** [comment]
> Here are other new assays that use antibodies:
> 
> ImmunoSABER
> Cell DIVE MxIF (cell DIVE is a commercial platform for hyperplaxed MxIF)
> MxIHC (Immuno-Histo-Compatibility)
> MIBI
> lightsheet

Cool: But none of them are in here right now, so I don't think any other action is required for now?
